hwcomposer/hwcomposer_backend_v11.cpp

```diff
@@ -86,122 +94,433 @@ static void hwc11_callback_invalidate(const struct hwc_procs *)
+    {
+        free(contents);
+
+        for (auto screen: screens) {
+            delete screen;
+        }
+    }
+
+    void update_screen_sizes(HwComposerBackend_v11 *backend)
+    {
+        for (auto screen: screens) {
+            screen->set_screen_size(backend->getSingleAttribute(HWC_DISPLAY_WIDTH, screen->get_id()), backend->getSingleAttribute(HWC_DISPLAY_HEIGHT, screen->get_id()));
+        }
+    }
+
+    void prepare(buffer_handle_t handle, int acquireFenceFd, int width, int height, bool geometryChanged) {
+        for (auto screen: screens) {
+            if (geometryChanged) {
+                screen->relayout(width, height);
+            }
+
+            screen->prepare(handle, acquireFenceFd, geometryChanged);
+        }
+
+        // We can close the acquire fence here, as we've dup'ed it for all displays here already
+        //sync_wait(acquireFenceFd, -1);
+        close(acquireFenceFd);
+    }
+
+    int release() {
+        int result = -1;
+
+        for (auto screen: screens) {
+            result = screen->release(result);
+        }
+
+        return result;
+    }
+
+    hwc_display_contents_1_t **get()
+    {
+        // Assemble content for all screens
+        for (int i=0; i<num_displays; i++) {
+            contents[i] = screens[i]->get();
+        }
+
+        return contents;
+    }
+
+private:
+    hwc_composer_device_1_t *hwc_device;
+    int num_displays;
+    std::vector<HwComposerScreen_v11 *> screens;
+    hwc_display_contents_1_t **contents;
+};
 
 HwComposerBackend_v11::HwComposerBackend_v11(hw_module_t *hwc_module, hw_device_t *hw_device, void *libminisf, int num_displays)
     : HwComposerBackend(hwc_module, libminisf)
     , hwc_device((hwc_composer_device_1_t *)hw_device)
-    , hwc_list(NULL)
     , hwc_mList(NULL)
     , num_displays(num_displays)
+    , m_screenAttachedGeometryChanged(true)
     , m_displayOff(true)
+    , width(0)
+    , height(0)
+    , content(new HwComposerContent_v11(hwc_device, num_displays))
 {
     procs = new HwcProcs_v11();
     procs->invalidate = hwc11_callback_invalidate;
     procs->hotplug = hwc11_callback_hotplug;
     procs->vsync = hwc11_callback_vsync;
     procs->backend = this;
+    procs->content = content;
 
     hwc_device->registerProcs(hwc_device, procs);
 
     hwc_version = interpreted_version(hw_device);
     sleepDisplay(false);
+    content->update_screen_sizes(this);
+}
+
+static void hwc11_callback_hotplug(const struct hwc_procs *procs, int disp, int connected)
+{
+    fprintf(stderr, "%s: procs=%x, disp=%d, connected=%d\n", __func__, procs, disp, connected);
+    if (disp == HWC_DISPLAY_EXTERNAL) {
+        g_external_connected_next = connected;
+    }
+
+    ((struct HwcProcs_v11*)procs)->backend->screenPlugged();
+    ((struct HwcProcs_v11*)procs)->content->update_screen_sizes(((struct HwcProcs_v11*)procs)->backend);
 }
 
 HwComposerBackend_v11::~HwComposerBackend_v11()
 {
     hwc_device->eventControl(hwc_device, 0, HWC_EVENT_VSYNC, 0);
```

Comment:
Should there be hwc_device->eventControl(hwc_device, 1, HWC_EVENT_VSYNC, 0) also, i.e. for the external display?